File: Agent.cpp
#include "Agent.h"
#include "Environment.h"
#include "SimpleBrain.h"
#include "MyError.h"
#include <stdlib.h>

void Agent::CalcFrontPos(int & x, int & y) {
	const unsigned siz = Environment::EnvSize;
	switch(direction) {
		case 0: x--; break; // u
		case 1: y++; break; // r
		case 2: x++; break; // d
		case 3: y--; break; // l
		default: MyError("Unknown direction.");
	}
	x = (x + siz) % siz;
	y = (y + siz) % siz;
}

Agent::Agent(int face, Brain * brn, int a) {
	appearance = face;
	health = MaxHealth;
	age = a;
	direction = rand() & 3;
	isAttacked = false;
	if(brn)
		brain = brn;
	else
		brain = new SimpleBrain(~0); // magic number: 0xFFFFF300
}

Agent::~Agent() {
	delete brain;
}

unsigned Agent::SeeBrain(void) const {
	return ((const SimpleBrain*)brain)->Show();
}

int Agent::MutateFace(void) const {
	return appearance ^ (1 << (rand() % FaceLength));
}

Brain * Agent::MutateBrain(void) const {
	return brain->Mutate();
}

inline bool IsPowerOfTwo(unsigned x) {
	return (x & (x - 1)) == 0;
}

inline bool IsSimilar(unsigned a, unsigned b) {
	return IsPowerOfTwo(a ^ b);
}

void Agent::Act(int x, int y, Environment * env) {
	const int grass = env->GetGrass(x, y);
	int x2 = x, y2 = y;
	CalcFrontPos(x2, y2);
	Agent * ag2 = env->GetAgent(x2, y2);
	bool b = brain->React(isAttacked, env->GetGrass(x2, y2) > grass, ag2 == NULL, ag2 && ag2->health > health,
		ag2 && IsSimilar(ag2->appearance, appearance));
	isAttacked = false;
	if(b){
		if(ag2){
			ag2->health -= grass;
			ag2->isAttacked = true;
			ag2->direction = direction ^ 2;
		}
		else{
			env->Move(x, y, x2, y2);
		}
	}
	else{
		direction = (direction + 1) & 3;
	}
}

File: agent.go
package main

type Direction uint8

const (
	Up Direction = iota
	Left
	Down
	Right
)

type Face uint8

type Agent struct {
	Health, Age int
	Face        Face
	IsAttacked  bool
	IsAttacking bool
	Brain       Brain
	Direction   Direction
}

func isPowerOfTwo(x uint8) bool {
	return (x & (x - 1)) == 0
}

func isSimilar(a, b Face) bool {
	return isPowerOfTwo(uint8(a) ^ uint8(b))
}

func relLoc(x, y, rx, ry int) (int, int) {
	return (x + rx + EnvSize) % EnvSize, (y + ry + EnvSize) % EnvSize
}

func nextLoc(x, y int, dir Direction) (int, int) {
	switch dir {
	case Up:
		return relLoc(x, y, 0, -1)
	case Left:
		return relLoc(x, y, -1, 0)
	case Down:
		return relLoc(x, y, 0, 1)
	case Right:
		return relLoc(x, y, 1, 0)
	default:
		return x, y
	}
}

func (ag *Agent) Act(x, y int, env *Environment) {
	x2, y2 := nextLoc(x, y, ag.Direction)
	ag2 := env.Cell[x2][y2].Agent
	input := uint32(0)
	if ag.IsAttacked {
		input += 16
	}
	if env.Cell[x2][y2].Food > env.Cell[x][y].Food {
		input += 8
	}
	if ag2 == nil {
		input += 4
	}
	if ag2 != nil && ag2.Health > ag.Health {
		input += 2
	}
	if ag2 != nil && isSimilar(ag2.Face, ag.Face) {
		input++
	}
	decision := (ag.Brain.react(input)) & 1
	ag.IsAttacked = false
	ag.IsAttacking = false
	if decision == 0 {
		ag.Direction = (ag.Direction + 1) & 3
	} else {
		if ag2 == nil {
			env.Cell[x2][y2].Agent = ag
			env.Cell[x][y].Agent = nil
		} else {
			ag2.Health -= int(1+env.Cell[x][y].Food)
			ag2.IsAttacked = true
			ag.IsAttacking = true
			ag2.Direction = ag.Direction ^ 2
			env.Aggressiveness[0]++
		}
	}
}